create_app factory with variable length keyword arguments raises an error

# Description

My `create_app` factory accepts optional variable length keyword arguments:

```
def create_app(**kwargs):
    pass
```

We use these to optionally inject pre-initialized service objects into the application, usually for testing purposes. This worked through Flask 1.1.4, but breaks due to changes regarding the `script_info` parameter made in 2.0.0.

```
Error: Detected factory 'create_app' in module 'service.main', but could not call it without arguments. Use "FLASK_APP='service.main:create_app(args)'" to specify arguments.
```

The buggy logic is in [call_factory](https://github.com/pallets/flask/blob/2.0.1/src/flask/cli.py#L106-L110); inspect considers the default value for a `**kwargs` argument to be empty. This causes the flask code to inject script_info as a positional argument, despite the function signature not accepting any positional args. A check needs to be added to exclude variable length kwargs, for example:

```
if (
    not args
    and len(sig.parameters) == 1
    and next(iter(sig.parameters.values())).default is inspect.Parameter.empty
    and next(iter(sig.parameters.values())).kind is not inspect.Parameter.VAR_KEYWORD
):  
```

# Test Case

Here's a test case that reproduces the bug:

```
class Module:
    @staticmethod
    def create_app(**kwargs):
        return Flask("appname")

app = find_best_app(script_info, Module)
assert isinstance(app, Flask)
assert app.name == "appname"
```

I was hoping to submit a simple patch for this issue, but the change I described above breaks tests for auto_reload and debug in ways that I don't quite understand.

```
src/flask/cli.py:75: NoAppException
___ test_flaskgroup_debug[True] ___

runner = <click.testing.CliRunner object at 0x10589f580>, set_debug_flag = True

    @pytest.mark.parametrize("set_debug_flag", (True, False))
    def test_flaskgroup_debug(runner, set_debug_flag):
        def create_app():
            app = Flask("flaskgroup")
            app.debug = True
            return app
    
        @click.group(cls=FlaskGroup, create_app=create_app, set_debug_flag=set_debug_flag)
        def cli(**params):
            pass
    
        @cli.command()
        def test():
            click.echo(str(current_app.debug))
    
        result = runner.invoke(cli, ["test"])
        assert result.exit_code == 0
>       assert result.output == f"{not set_debug_flag}\n"
E       AssertionError: assert 'True\n' == 'False\n'
E         - False
E         + True

tests/test_cli.py:406: AssertionError
___ test_templates_auto_reload_debug_run ___

app = <Flask 'flask_test'>, monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x1057ad9d0>

    def test_templates_auto_reload_debug_run(app, monkeypatch):
        def run_simple_mock(*args, **kwargs):
            pass
    
        monkeypatch.setattr(werkzeug.serving, "run_simple", run_simple_mock)
    
        app.run()
>       assert not app.templates_auto_reload
E       AssertionError: assert not True
E        +  where True = <Flask 'flask_test'>.templates_auto_reload

```

# Environment

- Python version: 3.8.6
- Flask version: 2.0.1

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -19,6 +19,8 @@ Unreleased
     :issue:`4157`
 -   Correctly handle raising deferred errors in CLI lazy loading.
     :issue:`4096`
+-   The CLI loader handles ``**kwargs`` in a ``create_app`` function.
+    :issue:`4170`
 
 
 Version 2.0.1

--- a/src/flask/cli.py
+++ b/src/flask/cli.py
@@ -103,18 +103,21 @@ def call_factory(script_info, app_factory, args=None, kwargs=None):
         )
         kwargs["script_info"] = script_info
 
-    if (
-        not args
-        and len(sig.parameters) == 1
-        and next(iter(sig.parameters.values())).default is inspect.Parameter.empty
-    ):
-        warnings.warn(
-            "Script info is deprecated and will not be passed as the"
-            " single argument to the app factory function in Flask"
-            " 2.1.",
-            DeprecationWarning,
-        )
-        args.append(script_info)
+    if not args and len(sig.parameters) == 1:
+        first_parameter = next(iter(sig.parameters.values()))
+
+        if (
+            first_parameter.default is inspect.Parameter.empty
+            # **kwargs is reported as an empty default, ignore it
+            and first_parameter.kind is not inspect.Parameter.VAR_KEYWORD
+        ):
+            warnings.warn(
+                "Script info is deprecated and will not be passed as the"
+                " single argument to the app factory function in Flask"
+                " 2.1.",
+                DeprecationWarning,
+            )
+            args.append(script_info)
 
     return app_factory(*args, **kwargs)
 

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -76,6 +76,15 @@ def test_find_best_app(test_apps):
 
     class Module:
         @staticmethod
+        def create_app(**kwargs):
+            return Flask("appname")
+
+    app = find_best_app(script_info, Module)
+    assert isinstance(app, Flask)
+    assert app.name == "appname"
+
+    class Module:
+        @staticmethod
         def create_app(foo):
             return Flask("appname")
 

